rhine-terminal/src/FRP/Rhine/Terminal.hs

```diff
@@ -0,0 +1,103 @@
+{- | Wrapper to write @terminal@ applications in Rhine, using concurrency.
+-}
+
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE RecordWildCards #-}
+module FRP.Rhine.Terminal
+  ( TerminalEventClock (..)
+  , flowTerminal
+  , terminalConcurrently
+  )
+  where
+
+-- base
+import Prelude hiding (putChar)
+import Control.Concurrent ()
+import Control.Concurrent.MVar ()
+import Data.IORef ()
+
+-- time
+import Data.Time.Clock ( getCurrentTime )
+
+-- terminal
+import System.Terminal
+    ( awaitEvent, runTerminalT, Event, Interrupt, TerminalT )
+import System.Terminal.Internal ( Terminal )
+
+-- transformers
+import Control.Monad.Trans.Reader
+
+-- rhine
+import FRP.Rhine.Clock.Proxy ()
+import FRP.Rhine
+import Control.Monad.Trans.Class (lift)
+
+-- * Terminal clock
+data TerminalEventClock = TerminalEventClock
+
+instance (Terminal t) => Clock (TerminalT t IO) TerminalEventClock
+  where
+    type Time TerminalEventClock = UTCTime
+    type Tag  TerminalEventClock = Either Interrupt Event
+
+    initClock TerminalEventClock = do
+      initialTime <- liftIO getCurrentTime
+      return
+        ( constM $ do
+            event <- awaitEvent
+            time <- liftIO getCurrentTime
+            return (time, event)
+        , initialTime
+        )
+
+instance GetClockProxy TerminalEventClock
+
+instance Semigroup TerminalEventClock where
+  t <> _ = t
+
+-- type TerminalClSF t m = ClSF (TerminalT t m) TerminalEventClock () (Either Interrupt Event)
+
+flowTerminal
+  :: ( Terminal t
+     , Clock (TerminalT t IO) cl
+     , GetClockProxy cl
+     , Time cl ~ Time (In  cl)
+     , Time cl ~ Time (Out cl)
+     )
+  => t
+  -> Rhine (TerminalT t IO) cl () ()
```

Comment:
Shouldn't this work for arbitrary inner monads other than `IO`?

Reply:
Yes, I'll transform this into a `MonadTerminal m` constraint when I figure out how to write `terminalConcurrently`

Reply:
I could write `flowTerminal` such that it returned an `m` with the `MonadIO m` constraint, but when I write it with the `MonadTerminal m` constraint:

```

flowTerminal'
  :: ( MonadIO m2
     , MonadMask m2
     , Terminal t
     , MonadTerminal m1
     , m1 ~ TerminalT t m2
     , Clock m1 cl
     , GetClockProxy cl
     , Time cl ~ Time (In  cl)
     , Time cl ~ Time (Out cl)
     )
  => t
  -> Rhine m1 cl () ()
  -> m2 ()
flowTerminal' term clsf = flip runTerminalT term $ flow clsf
```

It's not clear whether I'm gaining much and the `m1 ~ Terminal t m2` constraint seems suspect and I feel like I'm going in circles too :D